b2sdk/replication/monitoring.py

```diff
@@ -0,0 +1,281 @@
+
+
+@dataclass(frozen=True)
+class AbstractReplicationAttrs(metaclass=ABCMeta):
+    """
+    Some attributes of files which are meaningful
+    for replication monitoring and troubleshooting.
+    """
+
+    @abstractclassmethod
+    def from_files(
+        cls, source_file: Optional[B2Path], destination_file: Optional[B2Path]
+    ) -> 'AbstractReplicationAttrs':
+        pass
+
+
+@dataclass(frozen=True)
+class ReplicationAttrs(AbstractReplicationAttrs):
+    """
+    Some attributes of source and destination files and their relations
+    which are meaningful for replication monitoring and troubleshooting.
+    """
+
+    # source attrs
+    source_replication_status: Optional[ReplicationStatus] = None
+    source_has_hide_marker: Optional[bool] = None
+    source_has_sse_c_enabled: Optional[bool] = None
+    source_has_large_metadata: Optional[bool] = None
+    source_has_file_retention: Optional[bool] = None
+    source_has_legal_hold: Optional[bool] = None
+
+    # destination attrs
+    destination_replication_status: Optional[ReplicationStatus] = None
+
+    # source & destination relation attrs
+    metadata_differs: Optional[bool] = None
+
+    LARGE_METADATA_SIZE: ClassVar[int] = 2048
+
+    @classmethod
+    def from_files(
+        cls, source_file: Optional[B2Path], destination_file: Optional[B2Path]
+    ) -> 'ReplicationAttrs':
+        params = {}
+
+        if source_file:
+            source_file_version = source_file.selected_version
+            params.update(
+                {
+                    'source_replication_status':
+                        source_file_version.replication_status,
+                    'source_has_hide_marker':
+                        source_file.is_visible(),
+                    'source_has_sse_c_enabled':
+                        source_file_version.server_side_encryption.mode == EncryptionMode.SSE_C,
+                    'source_has_large_metadata':
+                        source_file_version.headers_size >= cls.LARGE_METADATA_SIZE,
+                    'source_has_file_retention':
+                        source_file_version.file_retention is not NO_RETENTION_FILE_SETTING,
+                    'source_has_legal_hold':
+                        source_file_version.legal_hold is not LegalHold.UNSET,
+                }
+            )
+
+        if destination_file:
+            params.update(
+                {
+                    'destination_replication_status':
+                        destination_file.selected_version.replication_status,
+                }
+            )
+
+        if source_file and destination_file:
+            params.update(
+                {
+                    'metadata_differs':
+                        source_file.selected_version.file_info !=
+                        destination_file.selected_version.file_info,
+                }
+            )
+
+        return cls(**params)
+
+
+@dataclass
+class AbstractReplicationReport(metaclass=ABCMeta):
```

Comment:
again this I think is `AbstractScannerReport` or something, should be moved to a separate module and then replication is one of the users of this. Maybe the first one, but still, we may have a different scanner that shows how many files of what size and content_type / file extension you have and that'll use the same abstract class then. Interesting idea btw, I think it could be easy to create, right?